Prevent DW2 frame register/unregister from getting mistakenly stripped

From: J.M. Eubank <john@thesnappy.net>

# HG changeset patch
# Parent  44e726802760a665e0a0de2f5406aa2ef22946b8
---
 gcc/config/i386/cygwin.h  |    6 +++++-
 gcc/config/i386/mingw32.h |    6 +++++-
 2 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/gcc/config/i386/cygwin.h b/gcc/config/i386/cygwin.h
index 73737f480..5bd9a65df 100644
--- a/gcc/config/i386/cygwin.h
+++ b/gcc/config/i386/cygwin.h
@@ -55,6 +55,10 @@ along with GCC; see the file COPYING3.  If not see
     fvtable-verify=std:vtv_end.o%s} \
    crtend.o%s"
 
+/* There is a bug when building i686 dw-2 exceptions
+   where gcc_s gets stripped which this works around */
+#define PREVENT_STRIP_REG_FRAME_INFO "--undefined=___deregister_frame_info --undefined=___register_frame_info"
+
 /* Normally, -lgcc is not needed since everything in it is in the DLL, but we
    want to allow things to be added to it when installing new versions of
    GCC without making a new CYGWIN.DLL, so we leave it.  Profiling is handled
@@ -65,7 +69,7 @@ along with GCC; see the file COPYING3.  If not see
  %{static|static-libgcc:-lgcc -lgcc_eh} \
  %{!static: \
    %{!static-libgcc: \
-     -lgcc_s -lgcc \
+     -lgcc_s " PREVENT_STRIP_REG_FRAME_INFO " -lgcc \
     } \
   } "
 #else
diff --git a/gcc/config/i386/mingw32.h b/gcc/config/i386/mingw32.h
index de9858873..4587dda95 100644
--- a/gcc/config/i386/mingw32.h
+++ b/gcc/config/i386/mingw32.h
@@ -145,6 +145,10 @@ along with GCC; see the file COPYING3.  If not see
   " LINK_SPEC_LARGE_ADDR_AWARE "\
   %(shared_libgcc_undefs)"
 
+/* There is a bug when building i686 dw-2 exceptions
+   where gcc_s gets stripped which this works around */
+#define PREVENT_STRIP_REG_FRAME_INFO "--undefined=___deregister_frame_info --undefined=___register_frame_info"
+
 /* Include in the mingw32 libraries with libgcc */
 #ifdef ENABLE_SHARED_LIBGCC
 #define SHARED_LIBGCC_SPEC " \
@@ -152,7 +156,7 @@ along with GCC; see the file COPYING3.  If not see
  %{!static: \
    %{!static-libgcc: \
      %{!shared-libgcc:-lgcc} \
-     %{shared-libgcc:-lgcc_s -lgcc} \
+     %{shared-libgcc:-lgcc_s " PREVENT_STRIP_REG_FRAME_INFO " -lgcc} \
     } \
   } "
 #else
